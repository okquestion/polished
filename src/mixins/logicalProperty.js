// @flow
import type { Style } from '../types/style'
import capitalizeString from '../internalHelpers/_capitalizeString'

const boxMap = {
  top: 'block-start',
  right: 'inline-end',
  bottom: 'block-end',
  left: 'inline-start',
}

const dimensionMap = {
  height: 'blockSize',
  width: 'inlineSize',
  minHeight: 'minBlockSize',
  maxHeight: 'maxBlockSize',
  minWidth: 'minInlineSize',
  maxWidth: 'maxInlineSize',
}

const baseMap = {
  horizontal: 'inline',
  vertical: 'block',
  right: 'end',
  left: 'start',
}

function camelcaseProperty(property: string) {
  const splitProperty = property.split('-')
  if (splitProperty.length > 1) {
    return splitProperty.reduce((acc, val) => `${acc}${capitalizeString(val)}`)
  }
  return property
}

export default function logicalProperty(
  property: string | Style,
  value?: string | number,
): Array<Style> {
  const camelcasedProperty = camelcaseProperty(property)

  let newProperty
  let newValue

  switch (camelcasedProperty) {
    case 'captionSide':
    case 'clear':
    case 'float':
      newProperty = camelcasedProperty
      newValue = boxMap[value]
      break
    case 'textAlign':
    case 'resize':
      newProperty = camelcasedProperty
      newValue = baseMap[value]
      break
    case 'height':
    case 'width':
      newProperty = dimensionMap[camelcasedProperty]
      newValue = value
      break
    case 'minHeight':
    case 'minWidth':
    case 'maxHeight':
    case 'maxWidth':
      newProperty = dimensionMap[camelcasedProperty]
      newValue = value
      break
    default:
      break
  }

  return [
    {
      [camelcasedProperty]: value,
    },
    {
      [newProperty]: newValue,
    },
  ]
}
