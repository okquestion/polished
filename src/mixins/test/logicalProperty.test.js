// @flow
import logicalProperty from '../logicalProperty'

describe('logicalProperty', () => {
  it('returns logical properties for caption-side', () => {
    expect(logicalProperty('captionSide', 'left')).toMatchSnapshot()
    expect(logicalProperty('captionSide', 'right')).toMatchSnapshot()
    expect(logicalProperty('caption-side', 'right')).toMatchSnapshot()
  })

  it('returns logical properties for float', () => {
    expect(logicalProperty('float', 'left')).toMatchSnapshot()
    expect(logicalProperty('float', 'right')).toMatchSnapshot()
  })

  it('returns logical properties for clear', () => {
    expect(logicalProperty('clear', 'left')).toMatchSnapshot()
    expect(logicalProperty('clear', 'right')).toMatchSnapshot()
  })

  it('returns logical properties for resize', () => {
    expect(logicalProperty('resize', 'horizontal')).toMatchSnapshot()
    expect(logicalProperty('resize', 'vertical')).toMatchSnapshot()
  })

  it('returns logical properties for text-align', () => {
    expect(logicalProperty('textAlign', 'left')).toMatchSnapshot()
    expect(logicalProperty('textAlign', 'right')).toMatchSnapshot()
    expect(logicalProperty('text-align', 'right')).toMatchSnapshot()
  })

  it('returns logical properties for height & width', () => {
    expect(logicalProperty('height', '200px')).toMatchSnapshot()
    expect(logicalProperty('width', '300px')).toMatchSnapshot()
  })

  it('returns logical properties for min/max-height & min/max- width', () => {
    expect(logicalProperty('minHeight', '200px')).toMatchSnapshot()
    expect(logicalProperty('minWidth', '300px')).toMatchSnapshot()
    expect(logicalProperty('maxHeight', '200px')).toMatchSnapshot()
    expect(logicalProperty('maxWidth', '300px')).toMatchSnapshot()
    expect(logicalProperty('min-height', '200px')).toMatchSnapshot()
    expect(logicalProperty('max-width', '300px')).toMatchSnapshot()
  })
})
