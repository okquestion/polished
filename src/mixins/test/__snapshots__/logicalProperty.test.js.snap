// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logicalProperty returns logical properties for caption-side 1`] = `
Array [
  Object {
    "captionSide": "left",
  },
  Object {
    "captionSide": "inline-start",
  },
]
`;

exports[`logicalProperty returns logical properties for caption-side 2`] = `
Array [
  Object {
    "captionSide": "right",
  },
  Object {
    "captionSide": "inline-end",
  },
]
`;

exports[`logicalProperty returns logical properties for caption-side 3`] = `
Array [
  Object {
    "captionSide": "right",
  },
  Object {
    "captionSide": "inline-end",
  },
]
`;

exports[`logicalProperty returns logical properties for clear 1`] = `
Array [
  Object {
    "clear": "left",
  },
  Object {
    "clear": "inline-start",
  },
]
`;

exports[`logicalProperty returns logical properties for clear 2`] = `
Array [
  Object {
    "clear": "right",
  },
  Object {
    "clear": "inline-end",
  },
]
`;

exports[`logicalProperty returns logical properties for float 1`] = `
Array [
  Object {
    "float": "left",
  },
  Object {
    "float": "inline-start",
  },
]
`;

exports[`logicalProperty returns logical properties for float 2`] = `
Array [
  Object {
    "float": "right",
  },
  Object {
    "float": "inline-end",
  },
]
`;

exports[`logicalProperty returns logical properties for height & width 1`] = `
Array [
  Object {
    "height": "200px",
  },
  Object {
    "blockSize": "200px",
  },
]
`;

exports[`logicalProperty returns logical properties for height & width 2`] = `
Array [
  Object {
    "width": "300px",
  },
  Object {
    "inlineSize": "300px",
  },
]
`;

exports[`logicalProperty returns logical properties for min/max-height & min/max- width 1`] = `
Array [
  Object {
    "minHeight": "200px",
  },
  Object {
    "minBlockSize": "200px",
  },
]
`;

exports[`logicalProperty returns logical properties for min/max-height & min/max- width 2`] = `
Array [
  Object {
    "minWidth": "300px",
  },
  Object {
    "minInlineSize": "300px",
  },
]
`;

exports[`logicalProperty returns logical properties for min/max-height & min/max- width 3`] = `
Array [
  Object {
    "maxHeight": "200px",
  },
  Object {
    "maxBlockSize": "200px",
  },
]
`;

exports[`logicalProperty returns logical properties for min/max-height & min/max- width 4`] = `
Array [
  Object {
    "maxWidth": "300px",
  },
  Object {
    "maxInlineSize": "300px",
  },
]
`;

exports[`logicalProperty returns logical properties for min/max-height & min/max- width 5`] = `
Array [
  Object {
    "minHeight": "200px",
  },
  Object {
    "minBlockSize": "200px",
  },
]
`;

exports[`logicalProperty returns logical properties for min/max-height & min/max- width 6`] = `
Array [
  Object {
    "maxWidth": "300px",
  },
  Object {
    "maxInlineSize": "300px",
  },
]
`;

exports[`logicalProperty returns logical properties for resize 1`] = `
Array [
  Object {
    "resize": "horizontal",
  },
  Object {
    "resize": "inline",
  },
]
`;

exports[`logicalProperty returns logical properties for resize 2`] = `
Array [
  Object {
    "resize": "vertical",
  },
  Object {
    "resize": "block",
  },
]
`;

exports[`logicalProperty returns logical properties for text-align 1`] = `
Array [
  Object {
    "textAlign": "left",
  },
  Object {
    "textAlign": "start",
  },
]
`;

exports[`logicalProperty returns logical properties for text-align 2`] = `
Array [
  Object {
    "textAlign": "right",
  },
  Object {
    "textAlign": "end",
  },
]
`;

exports[`logicalProperty returns logical properties for text-align 3`] = `
Array [
  Object {
    "textAlign": "right",
  },
  Object {
    "textAlign": "end",
  },
]
`;
